### PR TITLE
inital mcdropout code

### PR DESCRIPTION
**Pull request by hstojic**
initial MC dropout code with some basic tests, from the master thesis project 
idea is optimize and modify this code in further PRs after discussing pros/cons in this PR
I have started a project `mc_dropout` that we will populate with identified tickets

### CODE REVIEW

File: tests/integration/test_predictions.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
+# Copyright 2022 The deep_bayesopt Authors. All Rights Reserved.
```
Comment (willcowley) [inline tests/integration/test_predictions.py:1]:
```suggestion
# Copyright 2023 The deep_bayesopt Authors. All Rights Reserved.
```

File: tests/integration/test_predictions.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    stochastic_passes = tf.stack([model(inputs, training=True) for _ in range(num_passes)], axis=0)
+    predicted_means = tf.math.reduce_mean(stochastic_passes, axis=0)
+
+    predicted_vars = tf.subtract(
+        tf.divide(
+            tf.reduce_sum(tf.math.multiply(stochastic_passes, stochastic_passes), axis=0),
+            num_passes,
+        ),
+        tf.math.square(predicted_means),
+    )
```
Comment (willcowley) [inline tests/integration/test_predictions.py:122]: This will be a pain to do each time we want to make a prediction...I'm guessing we should add this to the `DropOutNetwork` (or some standardised `ua_net`?) api in the longer term...

File: tests/integration/test_predictions.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    y = tf.pow(x, 3)
+    if error:
+        y += tf.random.normal(x.shape, 0, 3, dtype=x.dtype)
+    return y
+
+
+def _plot_predictions(
+    x: tf.Tensor, y: tf.Tensor, mu: tf.Tensor, sigma2: tf.Tensor, name: str
+) -> None:
+    """plots 2 dimensional objective space with the front"""
```
Comment (willcowley) [inline tests/integration/test_predictions.py:37]: minor: is the docstring a relic from something else?

File: tests/integration/test_predictions.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    plt.figure(figsize=(7, 5))
+    # plot data
+    plt.scatter(x, y, marker="x", color="k", alpha=0.5)
+    # plot predictions
+    # lower = mu - tf.math.sqrt(sigma2)
+    # upper = mu + tf.math.sqrt(sigma2)
+    # plt.fill_between(tf.squeeze(x), tf.squeeze(lower), tf.squeeze(upper), color="C1", alpha=0.3)
```
Comment (willcowley) [inline tests/integration/test_predictions.py:44]: minor: do we not want to see these?

File: tests/uanets/models/test_mc_dropout.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+    # basics
+    assert isinstance(model, tf.keras.Model)
+
+    # check the model has not been compiled
+    assert model.compiled_loss is None
+    assert model.compiled_metrics is None
+    assert model.optimizer is None
+
+    # check correct number of layers and proerply constructed
```
Comment (willcowley) [inline tests/uanets/models/test_mc_dropout.py:75]:
```suggestion
    # check the number of layers is correct and they are properly constructed
```

File: tests/integration/test_predictions.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+            tf.reduce_sum(tf.math.multiply(stochastic_passes, stochastic_passes), axis=0),
+            num_passes,
+        ),
+        tf.math.square(predicted_means),
+    )
+    mean_abs_deviation = tf.reduce_mean(tf.abs(predicted_means - outputs))
+
+    # plotting data, accuracy and training loss
+    _plot_predictions(inputs, outputs, predicted_means, predicted_vars, "MC Dropout")
+    _plot_training_loss(model, "MC Dropout")
```
Comment (willcowley) [inline tests/integration/test_predictions.py:127]: I'm presuming saving these figures to `tmp` is for convenience atm?

**Review by willcowley — COMMENTED** [review]
looks good so far...only minor comments at this stage

File: uanets/models/mc_dropout.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    def _gen_hidden_layers(self) -> tf.keras.Model:
+
+        hidden_layers = tf.keras.Sequential(name="hidden_layers")
+        for hidden_layer_args in self._hidden_layer_args:
+            hidden_layers.add(
+                tf.keras.layers.Dropout(rate=self.rate, dtype=self.input_tensor_spec.dtype)
+            )
+            hidden_layers.add(
+                tf.keras.layers.Dense(dtype=self.input_tensor_spec.dtype, **hidden_layer_args)
+            )
```
[the remark this answers is not in the document]
Comment (willcowley) [reply uanets/models/mc_dropout.py:90]: Ah - I see the `Dropout` layer just deals with the dropout part and has no weight multiplication/activation function.

File: tests/integration/test_predictions.py (hunk 2 above)
Comment (willcowley) [reply tests/integration/test_predictions.py:122]: Ah - I see there's already a note/issue for this...

File: tests/integration/test_predictions.py (hunk 3 above)
Comment (hstojic) [reply tests/integration/test_predictions.py:37]: yeah, borrowed from another plot

File: tests/integration/test_predictions.py (hunk 4 above)
Comment (hstojic) [reply tests/integration/test_predictions.py:44]: we do, but the figure is not correct at the moment, didnt manage to get to that

Comment (hstojic) [reply tests/integration/test_predictions.py:44]: (my skills with matplotlib are bad...)

File: tests/integration/test_predictions.py (hunk 2 above)
Comment (hstojic) [reply tests/integration/test_predictions.py:122]: yes, thats why I thought we should have an interface with some methods - we can discuss which ones exactly would be necessary

File: tests/integration/test_predictions.py (hunk 6 above)
Comment (hstojic) [reply tests/integration/test_predictions.py:127]: these are tests, so not sure they should be saved anywhere else?

File: uanets/models/mc_dropout.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (hstojic) [reply uanets/models/mc_dropout.py:90]: yes...

File: tests/integration/test_predictions.py (hunk 3 above)
Comment (hstojic) [reply tests/integration/test_predictions.py:37]: addressed now

File: tests/uanets/models/test_mc_dropout.py (hunk 5 above)
Comment (hstojic) [reply tests/uanets/models/test_mc_dropout.py]: changed

File: tests/integration/test_predictions.py (hunk 2 above)
Comment (willcowley) [reply tests/integration/test_predictions.py:119]: cool, let's defer this discussion to #3

File: tests/integration/test_predictions.py (hunk 6 above)
Comment (willcowley) [reply tests/integration/test_predictions.py:124]: No, I guess my broader point was do we want to persist artefacts such as plots from the integration tests at all?  I can see how it's useful for checking things at this stage, however.

**Review by willcowley — APPROVED** [review]
LGTM!
